httputil/client_test.go

```diff
@@ -83,3 +100,115 @@ func (s *clientSuite) TestClientProxySetsUserAgent(c *check.C) {
 
 	c.Assert(called, check.Equals, true)
 }
+
+var privKey, _ = rsa.GenerateKey(rand.Reader, 768)
+
+// see crypto/tls/generate_cert.go
+func generateTestCert(c *check.C, certpath, keypath string) {
+	template := x509.Certificate{
+		SerialNumber: big.NewInt(123456789),
+		Subject: pkix.Name{
+			Organization: []string{"Snapd testers"},
+		},
+		NotBefore:   time.Now(),
+		NotAfter:    time.Now().Add(24 * time.Hour),
+		IPAddresses: []net.IP{net.IPv4(127, 0, 0, 1)},
+		DNSNames:    []string{"localhost"},
+		IsCA:        true,
+		KeyUsage:    x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+	}
+	derBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, &privKey.PublicKey, privKey)
+	c.Assert(err, check.IsNil)
+
+	certOut, err := os.Create(certpath)
+	c.Assert(err, check.IsNil)
+	err = pem.Encode(certOut, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
+	c.Assert(err, check.IsNil)
+	err = certOut.Close()
+	c.Assert(err, check.IsNil)
+
+	if keypath != "" {
+		keyOut, err := os.Create(keypath)
+		c.Assert(err, check.IsNil)
+		privBytes := x509.MarshalPKCS1PrivateKey(privKey)
+		err = pem.Encode(keyOut, &pem.Block{Type: "PRIVATE KEY", Bytes: privBytes})
+		c.Assert(err, check.IsNil)
+		err = keyOut.Close()
+		c.Assert(err, check.IsNil)
+	}
+}
+
+type tlsSuite struct {
+	testutil.BaseTest
+
+	tmpdir            string
+	certpath, keypath string
+	logbuf            *bytes.Buffer
+
+	srv *httptest.Server
+}
+
+var _ = check.Suite(&tlsSuite{})
+
+func (s *tlsSuite) SetUpTest(c *check.C) {
+	s.BaseTest.SetUpTest(c)
+
+	s.tmpdir = c.MkDir()
+	dirs.SetRootDir(s.tmpdir)
+	err := os.MkdirAll(dirs.SnapdExtraSSLCertsDir, 0755)
+	c.Assert(err, check.IsNil)
+
+	s.certpath = filepath.Join(dirs.SnapdExtraSSLCertsDir, "good.pem")
+	s.keypath = filepath.Join(c.MkDir(), "key.pem")
+	generateTestCert(c, s.certpath, s.keypath)
+
+	// create a server that uses our certs
+	s.srv = httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		io.WriteString(w, `all good`)
+	}))
+	cert, err := tls.LoadX509KeyPair(s.certpath, s.keypath)
+	c.Assert(err, check.IsNil)
+	s.srv.TLS = &tls.Config{Certificates: []tls.Certificate{cert}}
+	s.srv.StartTLS()
+	s.AddCleanup(s.srv.Close)
+
+	logbuf, restore := logger.MockLogger()
+	s.logbuf = logbuf
+	s.AddCleanup(restore)
+}
+
+func (s *tlsSuite) TestClientNoExtraSSLRefuses(c *check.C) {
+	// clear rootdir, no extra certs now
+	dirs.SetRootDir(c.MkDir())
+
+	// create a client, it should pick up our test cert
+	cli := httputil.NewHTTPClient(nil)
+	c.Assert(cli, check.NotNil)
+	c.Assert(s.logbuf.String(), check.Equals, "")
+
+	_, err := cli.Get(s.srv.URL)
+	c.Assert(err, check.ErrorMatches, ".* certificate signed by unknown authority")
+}
+
+func (s *tlsSuite) TestClientExtraSSLCertInvalidCertWarnsAndRefuses(c *check.C) {
+	err := ioutil.WriteFile(filepath.Join(dirs.SnapdExtraSSLCertsDir, "garbage.pem"), []byte("garbage"), 0644)
```

Comment:
A malformed pem is a good test, but how about a test that fails when `InsecureSkipVerify: false,` (ie, the default with, for example, a self-signed cert or when the host doesn't match the x509 names? This would ensure that if defaults changed or coding error caused `InsecureSkipVerify: true` (or similar) that we would catch the problem before it went into production.

Reply:
I added a test for this now.

Reply:
Did you forget to push it? I don't see a test for a well-formed but invalid cert or a test for a server with TLS 1.1.